nav2_controller/src/controller_server.cpp

```diff
@@ -607,6 +673,22 @@ void ControllerServer::updateGlobalPath()
       action_server_->terminate_current();
       return;
     }
+    std::string current_progress_checker;
+    if (findProgressCheckerId(goal->progress_checker_id, current_progress_checker)) {
+      if (current_progress_checker_ != current_progress_checker) {
+        RCLCPP_INFO(
+          get_logger(), "Change of progress checker %s requested, resetting it",
+          goal->progress_checker_id.c_str());
+        progress_checkers_[current_progress_checker]->reset();
+      }
+      current_progress_checker_ = current_progress_checker;
+    } else {
+      RCLCPP_INFO(
+        get_logger(), "Terminating action, invalid progress checker %s requested.",
+        goal->progress_checker_id.c_str());
+      action_server_->terminate_current();
+      return;
+    }
     setPlannerPath(goal->path);
```

Comment:
I think you need to `progress_checkers_[current_progress_checker_]->reset();` after `setPlannerPath()`. This is done in the main function on an action request to reset the state to a fresh start but not in preemption. This could have stale state from the last time this goal checker was used (or never, if never used)

Reply:
You only reset the _old_ one above, which I'm not necessarily convinced you need to do (?)

Reply:
Oh I see, you do `progress_checkers_[current_progress_checker]->reset();`, I read that the first time as `progress_checkers_[current_progress_checker_]->reset();`

Whereas the second is the old goal checker I thought you were resetting unnecessarily (but then not resetting the newly set one). That makes sense. That was my mistake. 

Why not move the `current_progress_checker_ = current_progress_checker;` into the nested if statement? I wouldn't do anything useful just within the first if statement when they are equal.

Reply:
> Why not move the `current_progress_checker_ = current_progress_checker;` into the nested if statement? I wouldn't do anything useful just within the first if statement when they are equal.

Agreed, I moved it in the if statement and now use **current_progress_checker_** for the reset, so it is clearer to read